refactor context

dbarrosop opened the pull request:
This is far from completed but wanted to open it in case people wanted to start looking into it.

Todo:

1. Review API changed and clean
2. Fix docstrings
3. Recover the Logger() functionality that gornir.Context had
4. Find a better name for TaskParameters

I like having an object like `TaskParameters` to try to encapsulate common parameters that need be passed around. Otherwise the signature grows quite large and require breaking changes if new parameters need to be added.

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -15,20 +15,25 @@ type Gornir struct {
+		&TaskParameters{
+			Title:  title,
+			Gornir: gr,
+			Logger: gr.Logger,
```
replace with something like `gr.Logger.WithNewID()` or similar that returns a new logger with an attached uid

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -44,10 +49,15 @@ func (g *Gornir) RunS(title string, runner Runner, task Task) (chan *JobResult,
+		&TaskParameters{
+			Title:  title,
+			Gornir: gr,
+			Logger: gr.Logger,
```
replace with something like gr.Logger.WithNewID() or similar that returns a new logger with an attached uid

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -44,10 +49,15 @@ func (g *Gornir) RunS(title string, runner Runner, task Task) (chan *JobResult,
+func (gr *Gornir) RunA(title string, runner Runner, task Task, results chan *JobResult) error {
 	err := runner.Run(
-		NewContext(context.Background(), title, g, g.Logger),
+		context.Background(), // TODO pass this?
```
we should probably let the user pass this to `RunA` to give more control to the user

nleiva commented on pkg/gornir/gornir.go:
```diff
@@ -44,10 +49,15 @@ func (g *Gornir) RunS(title string, runner Runner, task Task) (chan *JobResult,
+func (gr *Gornir) RunA(title string, runner Runner, task Task, results chan *JobResult) error {
 	err := runner.Run(
-		NewContext(context.Background(), title, g, g.Logger),
+		context.Background(), // TODO pass this?
```
I would pass context.TODO() for now (so is explicit that needs to be re-worked). I can then work on adding a cancellation mechanism to cancel all the child processes if the parent is terminated (to clean up all the goroutines created) . 

We would need to create the context before the Run command and then make sure we terminate the goroutines with the Done channel signal).

```go
ctx := context.Background()
ctx, cancel := context.WithCancel(ctx)
```

On the execution side:

```go
go func() {
	for {
		select {
		case <-ctx.Done():
			return // returning not to leak the goroutine
		...
		}
	}
}()
```

dbarrosop replied:
I agree

nleiva commented on examples/3_grouped_simple/main.go:
```diff
@@ -18,9 +19,9 @@ import (
+func (c *checkMemoryAndCPU) Run(ctx context.Context, wg *sync.WaitGroup, taskParameters *gornir.TaskParameters, jobResult chan *gornir.JobResult) {
 	// We instantiate a new object
-	result := gornir.NewJobResult(ctx)
+	result := gornir.NewJobResult(ctx, taskParameters)
```
I'd make the name of the arguments shorter. `tp *gornir.TaskParameters` instead of `taskParameters *gornir.TaskParameters`, etc. The scope of the function is limited so you see immediately what `tp` is for example.

```go
func (c *checkMemoryAndCPU) Run(ctx context.Context, wg *sync.WaitGroup, tp *gornir.TaskParameters, jr chan *gornir.JobResult) {
	// We instantiate a new object
	result := gornir.NewJobResult(ctx)
	result := gornir.NewJobResult(ctx, tp)
```

nleiva commented on pkg/gornir/gornir.go:
```diff
@@ -15,20 +15,25 @@ type Gornir struct {
 // This function will block until all the tasks are completed.
-func (g *Gornir) RunS(title string, runner Runner, task Task) (chan *JobResult, error) {
-	results := make(chan *JobResult, len(g.Inventory.Hosts))
+func (gr *Gornir) RunS(title string, runner Runner, task Task) (chan *JobResult, error) {
```
Why not call `RunS`, `RunSync` or else to make it easier for the reader?. Same thing for `RunA`...

dbarrosop replied:
No particular reason, happy to change.

nleiva commented on pkg/gornir/runner.go:
```diff
@@ -1,43 +1,68 @@
 	"sync"
 )
 
+type TaskParameters struct {
```
I'd try to separate the `hosts` and `task` parameters or not have the Host as part of this struct to decouple the two (will make testing simpler). Also is confusing to have a Host here, while Gornir has a list of Hosts as well.

nleiva replied:
I will add a separate issue to address the host portion, I think it will be better we reference an interface (behavior) instead  of a concrete type (implementation details).

dbarrosop replied:
> I'd try to separate the hosts and task parameters or not have the Host

The purpose of `TaskParameters` is twofold:

1. Simplify the signature and make things easier to pass around
2. Allow us to add new fields to it that may not be mandatory without breaking code

So having the host is convenient (even though I agree it's ugly-ish).

> is confusing to have a Host here, while Gornir has a list of Hosts as well

Yeah, thinking about it the Gornir object should go away from that structure.

> decouple the two (will make testing simpler)

Shouldn't be much different, in the end it's just a matter of calling a function either like `Run(ctx, host...)` or `Run(ctx, TaskParameters{host: host, ...}, ...)`

Overall I think this part is the trickiest from an API design perspective as I don't want to have a superlong/complex signature that may have to be changed otherwise but I agree combining "bare" arguments with `TaskParameters` is not the most elegant solution.

I think I need to think more about this.

As I think there is value in the decoupling of the context, worst case we could go with this solution (after addressing your other feedback) and open a new issue to discuss more about this.

dbarrosop commented:
More changes trying to address some of the comments:

1. Renamed `RunS/RunA` to `RunSync/RunAsync`
2. Renamed `TaskParameters` to `JobParameters` to distinguish between `Task` (the action to perform over the hosts) and a `Job` (the actual instance of executing a `Task` over a single `Host`).
2. `taskParameters` parameter has been renamed to `jp`
3. `*gornir.Gornir` is not passed around anymore instead:
   1. `FilterFunc` now receives a context.Context as first parameter instead of `*gornir.Gornir`
   2. Runners receive `map[string]*Host` directly

Still struggling with the signature of `Task.Run`, `JobParameters` and `JobResult` but unless there is something major I'd suggest going with this and keep discussing that in a different PR.

nleiva commented:
>    1. `FilterFunc` now receives a context.Context as first parameter instead of `*gornir.Gornir`

I don't get this part. You don't use/need the context for filtering the list of hosts, do you?.

dbarrosop commented:
> You don't use/need the context for filtering the list of hosts

No, that's there so it can be passed to the filter function in case the user wants to pass extra information via the context. We can discuss on a different issue if you think it's unnecessary.

> I would switch to value semantics

I guess you refer to the `*Hosts` part, that's basically because of #4. I am evaluating whether we could attach user defined data to the Hosts in which case tasks could populate/modify that data. This might be the wrong thing to do though so I'd suggest taking the discussion there
